File: 0623/solution.java

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            return new TreeNode(val, root, null);
        }
        traverse(root, val, depth, 1);
        return root;
    }

    private void traverse(TreeNode node, int val, int depth, int level) {
        if (node == null) {
            return;
        }

        if (level == depth - 1) {
            TreeNode left = node.left;
            TreeNode right = node.right;
            node.left = new TreeNode(val, left, null);
            node.right = new TreeNode(val, null, right);
        }

        else {
            traverse(node.left, val, depth, level + 1);
            traverse(node.right, val, depth, level + 1);
        }
    }
}

// time O(n)
// space O(depth)

File: 0623/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def addOneRow(
        self, root: Optional[TreeNode], val: int, depth: int
    ) -> Optional[TreeNode]:
        if depth == 1:
            return TreeNode(val, root, None)

        def traverse(node, val, level, depth):
            if not node:
                return
            if level == depth - 1:
                left, right = node.left, node.right
                node.left = TreeNode(val, left, None)
                node.right = TreeNode(val, None, right)
            else:
                traverse(node.left, val, level + 1, depth)
                traverse(node.right, val, level + 1, depth)

        traverse(root, val, 1, depth)
        return root
